contracts/ERC20MultiDelegate.sol

```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes.sol";
+
+/**
+ * @dev A child contract which will be deployed by the ERC20MultiDelegate utility contract
+ * This is a proxy delegator contract to vote given delegate on behalf of original delegator
+ */
+contract ERC20ProxyDelegator {
+    constructor(ERC20Votes _token, address _delegate) {
+        _token.approve(msg.sender, type(uint256).max);
+        _token.delegate(_delegate);
+    }
+}
+
+/**
+ * @dev A utility contract to let delegators to pick multiple delegate
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes public token;
+
+    /** ### EVENTS ### */
+
+    event ProxyDeployed(address indexed delegate, address proxyAddress);
+    event DelegationProcessed(
+        address indexed from,
+        address indexed to,
+        uint256 amount
+    );
+
+    /**
+     * @dev Constructor.
+     * @param _token The ERC20 token address
+     * @param _metadata_uri ERC1155 metadata uri
+     */
+    constructor(
+        ERC20Votes _token,
+        string memory _metadata_uri
+    ) ERC1155(_metadata_uri) {
+        token = _token;
+    }
+
+    /**
+     * @dev Executes the delegation transfer process for multiple source and target delegates.
+     * @param sources The list of source delegates.
+     * @param targets The list of target delegates.
+     * @param amounts The list of amounts to deposit/withdraw.
+     */
+    function delegateMulti(
+        address[] calldata sources,
+        address[] calldata targets,
+        uint256[] calldata amounts
+    ) external {
+        _delegateMulti(sources, targets, amounts);
+    }
+
+    function _delegateMulti(
+        address[] calldata sources,
```

Comment:
If you take `uint256` IDs instead of addresses, you could avoid the need to create the `sourceIds` and `targetIds` arrays, and save a lot of gas on that - just cast them back to addresses in the loop.